src/Halogen/Aff/Driver.purs

```diff
@@ -1,296 +1,98 @@
 module Halogen.Aff.Driver
-  ( RenderSpec
-  , runUI
+  ( runUI
+  , hydrateUI
+  , module Halogen.Aff.Driver.Implementation.Types
   , module Halogen
   ) where
 
 import Prelude
 
-import Control.Monad.Fork.Class (fork)
-import Control.Monad.Rec.Class (Step(..), tailRecM)
-import Control.Parallel (parSequence_)
-import Data.List ((:))
-import Data.List as L
-import Data.Map as M
-import Data.Maybe (Maybe(..), maybe, isJust, isNothing)
-import Data.Traversable (for_, sequence_, traverse_)
-import Data.Tuple (Tuple(..))
+import Data.Foldable (for_)
+import Data.Maybe (Maybe(..))
 import Effect (Effect)
-import Effect.Aff (Aff, killFiber)
+import Effect.Aff (Aff)
 import Effect.Class (liftEffect)
-import Effect.Console (warn)
-import Effect.Exception (error, throw)
 import Effect.Ref (Ref)
 import Effect.Ref as Ref
 import FRP.Event as Event
 import Halogen (HalogenIO)
 import Halogen.Aff.Driver.Eval as Eval
-import Halogen.Aff.Driver.State (DriverState(..), DriverStateRef(..), DriverStateX, LifecycleHandlers, RenderStateX, initDriverState, mapDriverState, renderStateX, renderStateX_, unDriverStateX)
-import Halogen.Component (Component, ComponentSlot, ComponentSlotBox, unComponent, unComponentSlot)
-import Halogen.Data.Slot as Slot
-import Halogen.HTML.Core as HC
-import Halogen.Query.HalogenQ as HQ
-import Halogen.Query.Input (Input)
-import Halogen.Query.Input as Input
+import Halogen.Aff.Driver.Implementation.Hydrate as Hydrate
```

Comment:
The `Implementation` modules may be a good idea in the future -- some refactoring certainly makes sense with this PR -- but in this pull request they make it difficult to see precisely what is changing. The best way to make sure the PR is reviewed is to make changes that produce the minimum diff necessary. That makes it easy on @garyb and @natefaubion to take a look and see exactly what's changed.

I understand why you'd want to do this refactoring as well, and once the main PR has been merged I think we'd all be fine with applying it afterwards.

Reply:
:( that would be hard for me to do